### PR TITLE
[cleanup] Deprecate more compat functions

### PR DESCRIPTION
**Pull request by seproDev**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

Some minor cleanup work

This PR right now does 3 things:
- Deprecate `compat_os_name`, `compat_realpath`, `compat_shlex_quote`, `bytes_to_intlist`, `compiled_regex_type`, and `intlist_to_bytes`
- Remove all usages of deprecated functions
- Move already long deprecated functions to legacy (emit warning when used)

Question: Should function that have been in legacy for more than a year be removed as part of this RP?

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

**Comment by grqz** [comment]
<details><summary>patch:

lets also deprecate `yt_dlp.utils.compiled_regex_type`
</summary>

```diff
diff --git a/pyproject.toml b/pyproject.toml
index 623121d76..a198be4a8 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -312,6 +312,7 @@ banned-from = [
 "yt_dlp.compat.compat_urllib_parse_urlencode".msg = "Use `urllib.parse.urlencode` instead."
 "yt_dlp.compat.compat_urllib_parse_urlparse".msg = "Use `urllib.parse.urlparse` instead."
 "yt_dlp.compat.compat_shlex_quote".msg = "Use `yt_dlp.utils.shell_quote` instead."
+"yt_dlp.utils.compiled_regex_type".msg = "Use `re.Pattern` instead."
 "yt_dlp.utils.error_to_compat_str".msg = "Use `str` instead."
 "yt_dlp.utils.bytes_to_intlist".msg = "Use `list` instead."
 "yt_dlp.utils.intlist_to_bytes".msg = "Use `bytes` instead."
diff --git a/yt_dlp/utils/_legacy.py b/yt_dlp/utils/_legacy.py
index d65b135d9..b64331fdd 100644
--- a/yt_dlp/utils/_legacy.py
+++ b/yt_dlp/utils/_legacy.py
@@ -16,6 +16,7 @@
 from ..networking._urllib import HTTPHandler
 
 # isort: split
+from re import Pattern as compiled_regex_type  # noqa: F401
 from .networking import escape_rfc3986  # noqa: F401
 from .networking import normalize_url as escape_url
 from .networking import random_user_agent, std_headers  # noqa: F401
diff --git a/yt_dlp/utils/_utils.py b/yt_dlp/utils/_utils.py
index 7fd0ed2a4..c5396789f 100644
--- a/yt_dlp/utils/_utils.py
+++ b/yt_dlp/utils/_utils.py
@@ -54,9 +54,6 @@
 
 __name__ = __name__.rsplit('.', 1)[0]  # noqa: A001: Pretend to be the parent module
 
-# This is not clearly defined otherwise
-compiled_regex_type = type(re.compile(''))
-
 
 class NO_DEFAULT:
     pass
```

</details>

**Comment by bashonly** [comment]
> 6b293bbeb368e50ade03d6fdff9f1c8a75c6d412 [Only warn for deprecated not legacy](https://github.com/yt-dlp/yt-dlp/pull/11439/commits/6b293bbeb368e50ade03d6fdff9f1c8a75c6d412)

IMO revert this. Devs should be warned for both via linter. Only legacy should warn during runtime

File: yt_dlp/extractor/adn.py [hunk 1]
```diff
@@ -198,7 +196,7 @@ def _real_extract(self, url):
 
         links_url = try_get(options, lambda x: x['video']['url']) or (video_base_url + 'link')
         self._K = ''.join(random.choices('0123456789abcdef', k=16))
-        message = bytes_to_intlist(json.dumps({
+        message = list(json.dumps({
             'k': self._K,
             't': token,
         }))
```
Comment (bashonly) [inline yt_dlp/extractor/adn.py:202]: this is wrong
```suggestion
        message = list(json.dumps({
            'k': self._K,
            't': token,
        }).encode())
```

File: yt_dlp/extractor/anvato.py [hunk 2]
```diff
@@ -277,8 +275,8 @@ def _get_video_json(self, access_key, video_id, extracted_token):
         server_time = self._server_time(access_key, video_id)
         input_data = f'{server_time}~{md5_text(video_data_url)}~{md5_text(server_time)}'
 
-        auth_secret = intlist_to_bytes(aes_encrypt(
-            bytes_to_intlist(input_data[:64]), bytes_to_intlist(self._AUTH_KEY)))
+        auth_secret = bytes(aes_encrypt(
+            list(input_data[:64]), list(self._AUTH_KEY)))
```
Comment (bashonly) [inline yt_dlp/extractor/anvato.py:279]: this is wrong, throws now

**Review by bashonly — CHANGES_REQUESTED** [review]

**Comment by bashonly** [comment]
This was just what I notcied, but I think all of the `bytes_to_intlist` and `intlist_to_bytes` replacements need to be scrutinized more closely

File: yt_dlp/extractor/anvato.py [hunk 3]
```diff
@@ -234,8 +232,8 @@ def _get_video_json(self, access_key, video_id, extracted_token):
         server_time = self._server_time(access_key, video_id)
         input_data = f'{server_time}~{md5_text(video_data_url)}~{md5_text(server_time)}'
 
-        auth_secret = intlist_to_bytes(aes_encrypt(
-            bytes_to_intlist(input_data[:64]), bytes_to_intlist(self._AUTH_KEY)))
+        auth_secret = bytes(aes_encrypt(
+            list(input_data[:64]).encode(), list(self._AUTH_KEY)))
```
Comment (bashonly) [inline yt_dlp/extractor/anvato.py:236]: still wrong, need to ecode the data being cast to list, not encode the list
```suggestion
            list(input_data[:64].encode()), list(self._AUTH_KEY)))
```

**Review by bashonly — APPROVED** [review]

**Review by bashonly — APPROVED** [review]

File: yt_dlp/extractor/shemaroome.py [hunk 4]
```diff
@@ -68,10 +66,10 @@ def _real_extract(self, url):
         data_json = self._download_json('https://www.shemaroome.com/users/user_all_lists', video_id, data=data.encode())
         if not data_json.get('status'):
             raise ExtractorError('Premium videos cannot be downloaded yet.', expected=True)
-        url_data = bytes_to_intlist(base64.b64decode(data_json['new_play_url']))
-        key = bytes_to_intlist(base64.b64decode(data_json['key']))
-        iv = [0] * 16
-        m3u8_url = unpad_pkcs7(intlist_to_bytes(aes_cbc_decrypt(url_data, key, iv))).decode('ascii')
+        url_data = base64.b64decode(data_json['new_play_url'])
+        key = base64.b64decode(data_json['key'])
+        iv = bytes([0] * 16)
```
Comment (bashonly) [inline yt_dlp/extractor/shemaroome.py:71]:
```suggestion
        iv = bytes(16)
```

**Review by bashonly — APPROVED** [review]

**Review by Grub4K — APPROVED** [review]

**Pull request merged**